march_joint_inertia_controller/include/march_joint_inertia_controller/joint_trajectory_inertia_controller.h

```diff
@@ -0,0 +1,214 @@
+                                           period);
+        inertia_estimators_[i].standard_deviation.push_back(inertia_estimators_[i].getAcceleration(0));
+        if (count_ == samples_)
+        {
+          inertia_estimators_[i].initP(samples_);
+        }
+      }
+    }
+    else
+    {
+      for (unsigned int i = 0; i < num_joints_; ++i)
+      {
+        // Update inertia estimator
+        inertia_estimators_[i].fillBuffers((*joint_handles_ptr_)[i].getVelocity(), (*joint_handles_ptr_)[i].getEffort(),
+                                           period);
+        inertia_estimators_[i].inertiaEstimate();
+        msg_.data = inertia_estimators_[i].getJointInertia();
+        pub_[i].publish(msg_);
+        // TO DO: Provide lookup table for gain selection
+        // TO DO: apply PID control
+
+        // Update PIDs
+        const double command = (pids_[i]->computeCommand(state_error.position[i], state_error.velocity[i], period));
+        (*joint_handles_ptr_)[i].setCommand(command);
+      }
+    }
+  }
+
+  /**
+   * \brief Starts the controller by checking if the joint handle pointer is filled, resets the pid calculators and sets
+   * the intial command to zero so that the joint doesn't start moving without a desired trajectory
+   */
+  void starting(const ros::Time& /*time*/)
+  {
+    if (!joint_handles_ptr_)
+    {
+      return;
+    }
+
+    // Reset PIDs, zero commands
+    for (unsigned int i = 0; i < pids_.size(); ++i)
+    {
+      pids_[i]->reset();
+      (*joint_handles_ptr_)[i].setCommand(0.0);
+    }
+  }
+
+  /**
+   * \brief No specific implementation needed I guess, but if necessary later then here it is.
+   */
+  void stopping(const ros::Time& /*time*/)
+  {
+  }
+
+  /**
+   * \brief Get the PID parameters
+   */
+  void getGains(int joint_nr, double& p, double& i, double& d, double& i_max, double& i_min)
+  {
+    bool dummy;
+    (*pids_[joint_nr]).getGains(p, i, d, i_max, i_min, dummy);
+  }
+
+  /**
+   * \brief Get the PID parameters with specification if anti windup is used
+   */
+  void getGains(int joint_nr, double& p, double& i, double& d, double& i_max, double& i_min, bool& antiwindup)
+  {
+    (*pids_[joint_nr]).getGains(p, i, d, i_max, i_min, antiwindup);
+  }
+
+  /**
+   * \brief Print debug info to console
+   */
+  void printDebug(int joint_nr)
+  {
+    (*pids_[joint_nr]).printValues();
+  }
+
+  /**
+   * \brief Set the PID parameters
+   */
+  void setGains(int joint_nr, const double& p, const double& i, const double& d, const double& i_max,
+                const double& i_min, const bool& antiwindup = false)
+  {
+    (*pids_[joint_nr]).setGains(p, i, d, i_max, i_min, antiwindup);
+  }
+
+private:
+  int count_ = 0;
+  int samples_;
+
+  typedef std::shared_ptr<control_toolbox::Pid> PidPtr;
+  std::vector<PidPtr> pids_;
+
+  std::vector<hardware_interface::JointHandle>* joint_handles_ptr_;
+
+  unsigned int num_joints_;
+
+  std::vector<ros::Publisher> pub_;
```

Comment:
You should make these publishers `realtime_tools::RealtimePublisher`. Because you publish inside of your update loop, which should be real time, these publishers can do that in real time. Take a look at how we use them in `march_hardware_interface.cpp`. You have even already included them at the top of this file, but you are not using them.

Reply:
Oeh, that's a good suggestion, I'll start on that ASAP.